hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ECUnderReplicationHandler.java

```diff
@@ -0,0 +1,242 @@
+        if (op.getOpType() == ContainerReplicaOp.PendingOpType.DELETE) {
+          deletionInFlight.add(op.getTarget());
+        }
+      }
+      List<Integer> missingIndexes = replicaCount.unavailableIndexes(true);
+      // We got the missing indexes, this is excluded any decommissioning
+      // indexes. Find the good source nodes.
+      if (missingIndexes.size() > 0 && !replicaCount.unRecoverable()) {
+        List<ContainerReplica> sources = replicas.stream().filter(r -> r
+            .getState() == StorageContainerDatanodeProtocolProtos
+            .ContainerReplicaProto.State.CLOSED)
+            // Exclude stale and dead nodes. This is particularly important for
+            // maintenance nodes, as the replicas will remain present in the
+            // container manager, even when they go dead.
+            .filter(r -> ReplicationManager
+                .getNodeStatus(r.getDatanodeDetails(), nodeManager).isHealthy())
+            .filter(r -> !deletionInFlight.contains(r.getDatanodeDetails()))
+            .collect(Collectors.toList());
+        LOG.debug("Missing indexes detected for the container {}." +
+                " The missing indexes are {}", id, missingIndexes);
+        // We have source nodes.
+        if (sources.size() >= repConfig.getData()) {
+          final List<DatanodeDetails> selectedDatanodes =
+              getTargetDatanodes(replicas, container, missingIndexes.size());
+
+          List<ReconstructECContainersCommand.DatanodeDetailsAndReplicaIndex>
+              sourceDatanodesWithIndex = new ArrayList<>();
+          for (ContainerReplica srcReplica : sources) {
+            sourceDatanodesWithIndex.add(
+                new ReconstructECContainersCommand
+                    .DatanodeDetailsAndReplicaIndex(
+                    srcReplica.getDatanodeDetails(),
+                    srcReplica.getReplicaIndex()));
+          }
+
+          final ReconstructECContainersCommand reconstructionCommand =
+              new ReconstructECContainersCommand(id.getProtobuf().getId(),
+                  sourceDatanodesWithIndex, selectedDatanodes,
+                  int2byte(missingIndexes),
+                  repConfig);
+          // Keeping the first target node as coordinator.
+          return ImmutableMap.of(selectedDatanodes.get(0),
+              reconstructionCommand);
+        } else {
+          LOG.warn("Cannot proceed for EC container reconstruction for {}, due"
+              + "to insufficient source replicas found. Number of source "
+              + "replicas needed: {}. Number of available source replicas are:"
+              + " {}. Available sources are: {}", container.containerID(),
+              repConfig.getData(), sources.size(), sources);
+        }
+      }
+
+      if (containerHealthResult.underReplicatedDueToDecommission()) {
+        Map<DatanodeDetails, SCMCommand<?>> commands = new HashMap<>();
+        Set<Integer> decomIndexes = replicaCount.decommissioningOnlyIndexes();
```

Comment:
This does not consider pendingAdd replicas - there may be some already scheduled for the decommission only indexes. I think this new `decommissioningOnlyIndexes` and its twin `maintenanceOnlyIndexes` need to take the `includePendingAdd` flag like I changed most of the other methods to do.

Reply:
Yeah, it's good count class is evolving by the use. Let's fix them and I will try to add test as well. 
includePendingAdd was completely ignored when we discuss offline in the morning. Thanks for catching. maintenanceOnlyIndexes also need that flag, but not in this patch as we have separate JIRA for that. I will add the comment there not to miss.